tests/trainer/test_trainer.py

```diff
@@ -491,6 +491,56 @@ def test_trainer_max_steps_and_epochs(tmpdir):
     assert trainer.global_step == num_train_samples * trainer.max_epochs
     assert trainer.current_epoch == trainer.max_epochs - 1, "Model did not stop at max_epochs"
 
+    # if max_steps is positive and max_epochs is negative, use max_steps
+    trainer_kwargs["max_epochs"] = -1
+    trainer_kwargs["max_steps"] = 3 * 2 * num_train_samples
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model)
+
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    assert trainer.global_step == 3 * 2 * num_train_samples
+
+
+@pytest.mark.parametrize(
+    "max_epochs,max_steps,incorrect_variable,incorrect_value",
+    [
+        (-100, None, "max_epochs", -100),
+        (1.5, None, "max_epochs", 1.5),
+        (1, -2, "max_steps", -2),
+        (1, 0.5, "max_steps", -2),
+    ],
+)
+def test_trainer_max_steps_and_epochs_validation(max_epochs, max_steps, incorrect_variable, incorrect_value):
+    """Don't allow max_epochs or max_steps to be less than -1 or a float"""
+    with pytest.raises(
+        MisconfigurationException,
+        match=f"`{incorrect_variable}` must be a positive integer or -1. You passed in {incorrect_value}",
+    ):
+        trainer = Trainer(max_epochs=max_epochs, max_steps=max_steps)
+
+
+@pytest.mark.parametrize(
+    "max_epochs,max_steps,is_done",
+    [
+        (None, None, False),
+        (-1, None, False),
+        (None, -1, False),
+        (5, -1, False),
+        (-1, 10, False),
+        (None, 0, True),
+        (0, None, True),
+        (-1, 0, True),
+        (0, -1, True),
+    ],
+)
+def test_trainer_max_steps_and_epochs_fit_loop_done(max_epochs, max_steps, is_done):
+    trainer = Trainer(max_epochs=max_epochs, max_steps=max_steps)
+
+    assert trainer.max_epochs == max_epochs
+    assert trainer.max_steps == max_steps
+    assert trainer.max_time is None
```

Comment:
have a look at the failing tests, this attribute does not exist on the trainer. However, to make your assert here you can check that there is no Timer() callback in the list of trainer.callbacks. 

To run individual tests locally, you can run

`py.test -v path/to/test_file.py::name_of_test_function
`
wasn't sure if you knew :)

Reply:
Thank you! That was so much easier!